File: matlab/examples/BenchMark.m
classdef BenchMark
  % [1] BenchMark.nativeTCPServer() + BenchMark.nativeTCPClient()

  % [2] BenchMark.xmatTCPServer(classname=double)  + BenchMark.xmatTCPClient(classname=double)

  properties (Constant)
    message_repeat = 32
    size_range = [10 20]  % [2^10 : 2^20]
    msg_size = 2.^(BenchMark.size_range(1):BenchMark.size_range(2));
    ip_port = xmat.TCPSocket.k_xport;
  end

  methods (Static)

    % native tcp-byte transfer
    % ------------------------
    function nativeTCPServer()
      fprintf('TestBench.nativeTCPServer\n');

      N = length(BenchMark.msg_size);

      socket = tcpserver("::", BenchMark.ip_port);

      t = zeros(N, BenchMark.message_repeat);
      legend_ = cell(1, N);
      for n = 1:N
        msg_size = BenchMark.msg_size(n);
        legend_{n} = sprintf('m:%12d', msg_size);
        fprintf('new message size: %d\n', msg_size);

        for m = 1:BenchMark.message_repeat
          t0 = tic();
          data = read(socket, msg_size, 'char');
          t(n, m) = toc(t0);
          fprintf('msg[%4d]\n', length(data));
        end

        fprintf('finish message size: %d Kb\n, time: %3.3f\n', msg_size/2^10, mean(t(n, 2:end)));
      end

      plot_result(t, 'native-Server', legend_);
    end

    function nativeTCPClient(ipaddress)
      fprintf('TestBench.nativeTCPClient\n');
      if nargin < 1
        ipaddress = "localhost";
      end

      N = length(BenchMark.msg_size);

      socket = tcpclient(ipaddress, BenchMark.ip_port);

      t = zeros(N, BenchMark.message_repeat);
      legend_ = cell(1, N);
      for n = 1:N
        msg_size = BenchMark.msg_size(n);
        legend_{n} = sprintf('m:%12d', msg_size);
        fprintf('new message size: %d\n', msg_size);

        for m = 1:BenchMark.message_repeat
          data = repmat(char('a' + (n - 1)), 1, msg_size);

          t0 = tic();
          write(socket, data, 'char');
          t(n, m) = toc(t0);
        end

        fprintf('finish message size: %d Kb, time: %3.3f\n', msg_size/2^10, mean(t(n, 2:end)));
      end

      plot_result(t, 'native-Client', legend_)
    end


    % xmat-format tcp-transfer: numeric elements
    % ------------------------------------------
    function xmatTCPServer()
      fprintf('TestBench.xmatTCPServer\n');

      N = length(BenchMark.msg_size);

      xtcp = xmat.TCPSocket.server('::', BenchMark.ip_port);

      t = zeros(N, BenchMark.message_repeat);
      legend_ = cell(1, N);
      for n = 1:N
        for m = 1:BenchMark.message_repeat
          t0 = tic();
          xin = xtcp.recv();
          data = xin.getitem('data');
          t(n, m) = toc(t0);
          sum_ = sum(data);
          fprintf('msg[%4d]\n', length(data));
        end
        msg_size = numel(data);
        legend_{n} = sprintf('m:%12d', msg_size);
        fprintf('new message size: %d\n', msg_size);

        fprintf('finish message size: %d K\n, time: %3.3f\n', msg_size/2^10, mean(t(n, 2:end)));
      end

      plot_result(t, 'xmat-Server', legend_);
    end

    function xmatTCPClient(dtype, ipaddress)
      % Parameters
      % ----------
      % classname: function_handle, {@double, @single, @complex, @int32, ...}
      %   numeric supported type

      fprintf('TestBench.xmatTCPClient\n');
      if nargin < 1
        dtype = @double;
      end
      if nargin < 2
        ipaddress = "localhost";
      end
      N = length(BenchMark.msg_size);

      xtcp = xmat.TCPSocket.client(ipaddress, BenchMark.ip_port);

      t = zeros(N, BenchMark.message_repeat);
      legend_ = cell(1, N);
      for n = 1:N
        msg_size = BenchMark.msg_size(n);
        legend_{n} = sprintf('m:%12d', msg_size);
        fprintf('new message size: %d\n', msg_size);

        data = dtype(n:n+msg_size);
        for m = 1:BenchMark.message_repeat
          t0 = tic();
          xout = xmat.MapStreamOut.byte();
          xout.setitem('data', data);
          xout.close();
          xtcp.send(xout);
          t(n, m) = toc(t0);
        end

        fprintf('finish message size: %d K\n, time: %3.3f\n', msg_size/2^10, mean(t(n, 2:end)));
      end

      plot_result(t, 'xmat-Client', legend_);
    end
  end
end


function plot_result(t, name, legend_)
fig = figure('Name', name);
subplot(1, 2, 1);
plot(t(:, 2:end).')
legend(legend_)
grid('on')

subplot(1, 2, 2);
plot(mean(t(:, 2:end), 2))
grid('on')
title('avarage time, sec')
xlabel('message length, 2\^(10 + x - 1) elements');

% summary

end

File: python/xmat/benchmark.py
import time
from importlib.util import find_spec

import numpy as np
from . import datastream
from . import tcpsocket

_MESSAGE_REPEAT = 32
_SIZE_RANGE = [10, 20]
_MSG_SIZE = tuple(2**k for k in range(_SIZE_RANGE[0], _SIZE_RANGE[1] + 1))
_IP_PORT = tcpsocket.PORT


def native_tcp_server():
	print('benchmark.native_tcp_server')
	N = len(_MSG_SIZE)
	listener = tcpsocket.TCPListener((tcpsocket.IPHOST, _IP_PORT), 1)
	connection = listener.accept()

	t = np.zeros((N, _MESSAGE_REPEAT), dtype=float)
	legend_ = []
	for n in range(N):
		msg_size = _MSG_SIZE[n]
		legend_.append(f'm:{msg_size:12d}')
		print('new message size: ', msg_size)
		for m in range(_MESSAGE_REPEAT):
			t0 = time.time()
			data = connection.recv_(msg_size)
			t[n, m] = time.time() - t0
			#  print content just for verification
			print('msg: ', len(data), 'data[-8::]', data[-8::])

		print(f'finish message size: {msg_size/2**10} Kb\n, time: {np.mean(t[n, 2:]):3.6f}\n')
	plot_result(t, 'native-server', legend_)


def native_tcp_client(ipaddress: str = tcpsocket.IPLOCALHOST):
	print('benchmark.native_tcp_client')
	N = len(_MSG_SIZE)
	datatotal = bytes(n % 256 for n in range(_MSG_SIZE[-1] + _MESSAGE_REPEAT + N))
	print('datatotal is done:')

	connection = tcpsocket.TCPConnection((ipaddress, _IP_PORT))
	t = np.zeros((N, _MESSAGE_REPEAT), dtype=float)
	legend_ = []
	for n in range(N):
		msg_size = _MSG_SIZE[n]
		legend_.append(f'm:{msg_size:12d}')
		print('new message size: ', msg_size)
		for m in range(_MESSAGE_REPEAT):
			t0 = time.time()
			connection.send_(datatotal[n+m:n+m+msg_size])
			t[n, m] = time.time() - t0
			#  print content just for verification
			print('data[-8::]', datatotal[n+m+msg_size-8:n+m+msg_size])

		print(f'finish message size: {msg_size/2**10} Kb\n, time: {np.mean(t[n, 2:]):3.6f}\n')
	plot_result(t, 'native-client', legend_)


# xmat - format tcp - transfer: numeric elements
# ----------------------------------------------
def xmat_tcp_server():
	print('benchmark.native_tcp_server')
	N = len(_MSG_SIZE)
	listener = tcpsocket.TCPListener((tcpsocket.IPHOST, _IP_PORT), 1)
	connection = listener.accept()

	t = np.zeros((N, _MESSAGE_REPEAT), dtype=float)
	legend_ = []
	for n in range(N):
		msg_size = _MSG_SIZE[n]
		legend_.append(f'm:{msg_size:12d}')
		print('new message size: ', msg_size)
		for m in range(_MESSAGE_REPEAT):
			t0 = time.time()
			xin = connection.recv()
			data = xin.getitem('data')
			t[n, m] = time.time() - t0
			#  print content just for verification
			print('msg: ', len(data), 'data[-8::]', data[-8::])

		print(f'finish message size: {msg_size/2**10} K\n, time: {np.mean(t[n, 2:]):3.6f}\n')
	plot_result(t, 'native-server', legend_)


def xmat_tcp_client(dtype: str = 'float', ipaddress: str = tcpsocket.IPLOCALHOST):
	print('benchmark.xmat_tcp_client')
	N = len(_MSG_SIZE)
	# datatotal = bytes(n % 256 for n in range(_MSG_SIZE[-1] + _MESSAGE_REPEAT + N))
	datatotal = np.arange(_MSG_SIZE[-1] + _MESSAGE_REPEAT + N, dtype=dtype)

	print('datatotal is done:')

	connection = tcpsocket.TCPConnection((ipaddress, _IP_PORT))
	t = np.zeros((N, _MESSAGE_REPEAT), dtype=float)
	legend_ = []
	for n in range(N):
		msg_size = _MSG_SIZE[n]
		legend_.append(f'm:{msg_size:12d}')
		print('new message size: ', msg_size)
		for m in range(_MESSAGE_REPEAT):
			t0 = time.time()
			xout = datastream.MapStreamOut.byte()
			xout.setitem('data', datatotal[n + m:n + m + msg_size])
			xout.close()
			connection.send(xout)
			t[n, m] = time.time() - t0
			#  print content just for verification
			print('data[-8::]', datatotal[n + m + msg_size - 8:n + m + msg_size])

		print(f'finish message size: {msg_size/2**10} K\n, time: {np.mean(t[n, 2:]):3.6f}\n')
	plot_result(t, 'native-client:' + dtype, legend_)


def plot_result(t, figure_name, legend):
	s = figure_name + ' summary: \n'
	ss = [f'{_MSG_SIZE[n]/2**10: 12.0f}K: {np.mean(t[n, 2:]): 4.9f} sec\n' for n in range(t.shape[0])]
	print(''.join([s] + ss))

	if not find_spec('matplotlib'):
		return
	import matplotlib.pyplot as plt
	fig = plt.figure(figure_name)
	ax = fig.add_axes((0.1, 0.1, 0.8, 0.8))
	ax.plot(t[2:, :].T, marker='o')
	ax.legend(legend)
	ax.grid(True)
	plt.show()
